### ISSUE TITLE
npcap 0.86 service fails to create - nmap 7.40 cannot scan

### ISSUE THREAD

**Issue by tgfruth1** [nmap/nmap#875]
I have encountered this issue on multiple Windows 2012R2 VMware VM's in our environment.  They are about as close to identical to each other as I can easily ascertain, meaning they have the same Windows updates, the same virtual hardware, the same Visual C++ runtime versions, etc.  I have gone multiple rounds with npcap 0.86.  I thought the failure to install the npcap service was due to DLL issue that was resolved by installing the necessary Visual C++ runtime.  However, on one machine in particular, this has not resolved the error. I have tried a reset of the IP stack.  I have tried manually scrubbing the registry and file system, rebooting, and reinstalling npcap, only for it to fail.  Unfortunately, I am not currently seeing useful information in the install log. I want to run the latest npcap along with the latest nmap.  I have succeeded in getting this installed on 6 of 7 systems, most of them Windows 2012R2. I'm stuck on this last one.

Trying from the administrative command line:

C:\Program Files\Npcap>NPFInstall.exe -i
Unknown error! e0000247
Npcap LWF driver has failed to be installed.

C:\Program Files\Npcap>NPFInstall.exe -iw
Npcap WFP callout driver has been successfully installed!

C:\Program Files\Npcap>NPFInstall.exe -i
Unknown error! e0000247
Npcap LWF driver has failed to be installed.

C:\Program Files\Npcap>NPFInstall.exe -check_dll
<NULL>

C:\Program Files\Npcap>NPFInstall.exe -kill_proc
All the processes that are still using Npcap DLLs have been successfully termina
ted!

C:\Program Files\Npcap>NPFInstall.exe -c
Npcap driver cache in Driver Store has been successfully cleaned up!

C:\Program Files\Npcap>NPFInstall.exe -i
Unknown error! e0000247
Npcap LWF driver has failed to be installed.

C:\Program Files\Npcap>NPFInstall.exe -d
Npcap service is not pending to stop.

C:\Program Files\Npcap>NPFInstall.exe -r (resulted in APPCRASH)

![image](https://cloud.githubusercontent.com/assets/28453868/25758213/3744a4aa-3193-11e7-9797-690681e2e43d.png)

[DiagReport-20170505-175412.txt](https://github.com/nmap/nmap/files/979902/DiagReport-20170505-175412.txt)

[install.txt](https://github.com/nmap/nmap/files/979903/install.txt)
[NPFInstall.txt](https://github.com/nmap/nmap/files/979904/NPFInstall.txt)

**Comment by hsluoyz** [comment]
From your NPFInstall.txt, your error is: ``SetupCopyOEMInfW: error, errCode = 0xe0000247.``. So the driver fails to install.

Please provide another file: ``C:\Windows\INF\setupapi.dev.log``. This file can be big. If you don't want to upload the whole file, you can search ``NPFInstall.exe`` in this file and find the Npcap-related section and provide that section only. It has a timestamp so it's not hard to identify.

**Comment by tgfruth1** [comment]
I have uploaded all the files that seemed relevant based on the file name you have provided.



[setupapi.zip](https://github.com/nmap/nmap/files/983960/setupapi.zip)

**Comment by hsluoyz** [comment]
```
>>>  [SetupCopyOEMInf - C:\Program Files\Npcap\NPCAP.inf]
>>>  Section start 2017/05/05 18:00:22.642
      cmd: NPFInstall.exe  -i
     sto: {Setup Import Driver Package: C:\Program Files\Npcap\NPCAP.inf} 18:00:22.642
     inf:      Provider: Nmap Project
     inf:      Class GUID: {4D36E974-E325-11CE-BFC1-08002BE10318}
     inf:      Driver Version: 04/12/2017,8.10.36.209
     inf:      Catalog File: npcap.cat
     sto:      {Copy Driver Package: C:\Program Files\Npcap\NPCAP.inf} 18:00:22.642
     sto:           Driver Package = C:\Program Files\Npcap\NPCAP.inf
     sto:           Flags          = 0x00000007
     sto:           Destination    = C:\Users\%my_user_name%\AppData\Local\Temp\{18167aae-7d34-2248-9e94-8207df61c87e}
     sto:           Copying driver package files to 'C:\Users\%my_user_name%\AppData\Local\Temp\{18167aae-7d34-2248-9e94-8207df61c87e}'.
     flq:           Copying 'C:\Program Files\Npcap\npcap.cat' to 'C:\Users\%my_user_name%\AppData\Local\Temp\{18167aae-7d34-2248-9e94-8207df61c87e}\npcap.cat'.
     flq:           Copying 'C:\Program Files\Npcap\NPCAP.inf' to 'C:\Users\%my_user_name%\AppData\Local\Temp\{18167aae-7d34-2248-9e94-8207df61c87e}\NPCAP.inf'.
     flq:           Copying 'C:\Program Files\Npcap\npcap.sys' to 'C:\Users\%my_user_name%\AppData\Local\Temp\{18167aae-7d34-2248-9e94-8207df61c87e}\npcap.sys'.
     sto:      {Copy Driver Package: exit(0x00000000)} 18:00:22.720
     pol:      {Driver package policy check} 18:00:22.736
     pol:      {Driver package policy check - exit(0x00000000)} 18:00:22.736
     sto:      {Stage Driver Package: C:\Users\%my_user_name%\AppData\Local\Temp\{18167aae-7d34-2248-9e94-8207df61c87e}\NPCAP.inf} 18:00:22.736
     inf:           {Query Configurability: C:\Users\%my_user_name%\AppData\Local\Temp\{18167aae-7d34-2248-9e94-8207df61c87e}\NPCAP.inf} 18:00:22.736
     inf:                Driver package 'NPCAP.inf' is configurable.
     inf:           {Query Configurability: exit(0x00000000)} 18:00:22.736
     flq:           Copying 'C:\Users\%my_user_name%\AppData\Local\Temp\{18167aae-7d34-2248-9e94-8207df61c87e}\npcap.cat' to 'C:\Windows\System32\DriverStore\Temp\{22d02799-1139-6a4a-b861-c11c5e4df648}\npcap.cat'.
     flq:           Copying 'C:\Users\%my_user_name%\AppData\Local\Temp\{18167aae-7d34-2248-9e94-8207df61c87e}\NPCAP.inf' to 'C:\Windows\System32\DriverStore\Temp\{22d02799-1139-6a4a-b861-c11c5e4df648}\NPCAP.inf'.
     flq:           Copying 'C:\Users\%my_user_name%\AppData\Local\Temp\{18167aae-7d34-2248-9e94-8207df61c87e}\npcap.sys' to 'C:\Windows\System32\DriverStore\Temp\{22d02799-1139-6a4a-b861-c11c5e4df648}\npcap.sys'.
     sto:           {DRIVERSTORE IMPORT VALIDATE} 18:00:22.814
     sig:                {_VERIFY_FILE_SIGNATURE} 18:00:22.830
     sig:                     Key      = NPCAP.inf
     sig:                     FilePath = C:\Windows\System32\DriverStore\Temp\{22d02799-1139-6a4a-b861-c11c5e4df648}\NPCAP.inf
     sig:                     Catalog  = C:\Windows\System32\DriverStore\Temp\{22d02799-1139-6a4a-b861-c11c5e4df648}\npcap.cat
!    sig:                     Verifying file against specific (valid) catalog failed! (0x800b0109)
!    sig:                     Error 0x800b0109: A certificate chain processed, but terminated in a root certificate which is not trusted by the trust provider.
     sig:                {_VERIFY_FILE_SIGNATURE exit(0x800b0109)} 18:00:22.861
     sig:                {_VERIFY_FILE_SIGNATURE} 18:00:22.861
     sig:                     Key      = NPCAP.inf
     sig:                     FilePath = C:\Windows\System32\DriverStore\Temp\{22d02799-1139-6a4a-b861-c11c5e4df648}\NPCAP.inf
     sig:                     Catalog  = C:\Windows\System32\DriverStore\Temp\{22d02799-1139-6a4a-b861-c11c5e4df648}\npcap.cat
!    sig:                     Verifying file against specific Authenticode(tm) catalog failed! (0x800b010a)
!    sig:                     Error 0x800b010a: A certificate chain could not be built to a trusted root authority.
     sig:                {_VERIFY_FILE_SIGNATURE exit(0x800b010a)} 18:00:22.861
!!!  sig:                Driver package catalog file certificate does not chain to a root, and Code Integrity is enforced.
!!!  sig:                Driver package failed signature validation. Error = 0xE0000247
     sto:           {DRIVERSTORE IMPORT VALIDATE: exit(0xe0000247)} 18:00:22.861
!!!  sig:           Driver package failed signature verification. Error = 0xE0000247
!!!  sto:           Failed to import driver package into Driver Store. Error = 0xE0000247
     sto:      {Stage Driver Package: exit(0xe0000247)} 18:00:22.877
     sto: {Setup Import Driver Package - exit (0xe0000247)} 18:00:22.877
!!!  inf: Failed to import driver package into driver store
!!!  inf: Error 0xe0000247: A problem was encountered while attempting to add the driver to the store.
<<<  Section end 2017/05/05 18:00:22.986
<<<  [Exit status: FAILURE(0xe0000247)]
```

The above content seems to say that the root certificate used by our EV cert is not trusted by Microsoft. I don't know why this happens. Maybe your machine missed some important security updates. Please update all KBs and retry the installation.

Also, you can find the ``npcap.sys`` in the Npcap install directory. Check its cert in its properties window like below. Check every cert in the cert tree to see if any cert in the chain is untrusted (maybe has a red cross thing or something like that). Then manually install and trust that untrusted cert.

![image](https://cloud.githubusercontent.com/assets/3787410/25811793/f7fe0744-3446-11e7-8673-0b0e8b9eea5f.png)

If this doesn't help, the last resort would be disabling the driver signature enforcement like [this](https://blogs.msdn.microsoft.com/emberger/2014/04/21/windows-server-2012-r2-install-unsigned-drivers/). This is a little insecure, but it counts as a workaround.

**Comment by tgfruth1** [comment]
The server in question appears to be up to date, and has no hidden updates that have not been installed. Is there a specific KB that applies?

**Comment by hsluoyz** [comment]
No. Then you should go Method 2 & Method 3.

**Comment by tgfruth1** [comment]
Properties of Digital Signatures is different.

![image](https://cloud.githubusercontent.com/assets/28453868/25813378/ec180348-33de-11e7-9e42-80bdbe67a690.png)

**Comment by tgfruth1** [comment]
From my perspective, based on your screenshots and mine, the Microsoft cert was left out of the 0.86 version of npcap.sys.  Here's a screenshot of 0.78-r5.

![image](https://cloud.githubusercontent.com/assets/28453868/25816513/ed05ad7c-33e9-11e7-8219-87f4c4c201f9.png)

**Comment by tgfruth1** [comment]
Just so you know, tried manually installing the Digicert certificate chain, no luck. Disabled driver signing, and I'm installing npcap 0.86.

![image](https://cloud.githubusercontent.com/assets/28453868/25817602/77a5e64c-33ed-11e7-85fa-ed1c0bcc48aa.png)

**Comment by tgfruth1** [comment]
My colleague says that https://www.digicert.com/CACerts/DigiCertAssuredIDRootCA.crt must beinstalled to fix issue with the code signing CA.

**Comment by hsluoyz** [comment]
So your system is fine with 0.78 r5? If yes, then please try all versions between 0.78 r5 and 0.86 and find the version that starts to fail. This will help our debugging here. Thanks.

**Comment by tgfruth1** [comment]
Installs from 0.78r3 thru 0.80 succeed.  0.81 is first to show the error:

![image](https://cloud.githubusercontent.com/assets/28453868/26309695/98d2bcce-3ec3-11e7-82e0-70d5b8ec57e3.png)

Any coincidence this is the first version downloaded from the new hosting location?

"Npcap 0.81

@dmiller-nmap dmiller-nmap released this on Feb 16 · 23 commits to master since this release

    Moved distribution of executable installer to http://npcap.org/ from Github."

**Comment by chikara1005** [comment]
I'm experiencing the same issues with NMAP version 7.40, and I've tried using npcap-0.86.  I've seen the following error in setupapi.dev.log, "Error 0x800b0109: A certificate chain processed, but terminated in a root certificate which is not trusted by the trust provider", and I've checked that I have the correct DigiCert Certificate installed.  The version that tgfruth1 stated was already installed on my system.  Do we have any solutions for .86, or should we just be running .80?  I'm new to all of this, so forgive me if that question seems ridiculous.

**Comment by chikara1005** [comment]
Sorry for being such a NEWB, but I went through the install again.  I'm using npcap for NMAP.  I don't believe I need the WinPcap for NMAP, so I removed the WinPcap Compatibility from the install, and everything went through fine.  Hope that helps.

**Comment by dmiller-nmap** [comment]
The "Microsoft Windows Hardware Compatibility Publisher" signature should only be present on Windows 10. It is added by Microsoft during the Attestation Signing process. In our testing, drivers signed like this are not able to be used on other versions of Windows; otherwise we would be able to reduce the number of drivers bundled in the installer.

It appears that one of the signatures is missing from the `npcap.cat` file. I will try to figure out why this is happening:

```
>signtool verify /all /v /kp win8_below\x64\npcap.cat

Verifying: win8_below\x64\npcap.cat
Signature Index: 0 (Primary Signature)
Hash of file (sha1): 806EDDE99F950BC9E7906444671CA9966E19367A

Signing Certificate Chain:
    Issued to: DigiCert High Assurance EV Root CA
    Issued by: DigiCert High Assurance EV Root CA
    Expires:   Sun Nov 09 19:00:00 2031
    SHA1 hash: 5FB7EE0633E259DBAD0C4C9AE6D38F1A61C7DC25

        Issued to: DigiCert High Assurance Code Signing CA-1
        Issued by: DigiCert High Assurance EV Root CA
        Expires:   Tue Feb 10 07:00:00 2026
        SHA1 hash: E308F829DC77E80AF15EDD4151EA47C59399AB46

            Issued to: Insecure.Com LLC
            Issued by: DigiCert High Assurance Code Signing CA-1
            Expires:   Wed Jun 27 07:00:00 2018
            SHA1 hash: 67CDCA7703A01B25E6E0426072EC08B0046EB5F8

The signature is timestamped: Fri May 26 17:21:45 2017
Timestamp Verified by:
    Issued to: DigiCert Assured ID Root CA
    Issued by: DigiCert Assured ID Root CA
    Expires:   Sun Nov 09 19:00:00 2031
    SHA1 hash: 0563B8630D62D75ABBC8AB1E4BDFB5A899B24D43

        Issued to: DigiCert Assured ID CA-1
        Issued by: DigiCert Assured ID Root CA
        Expires:   Tue Nov 09 19:00:00 2021
        SHA1 hash: 19A09B5A36F4DD99727DF783C17A51231A56C117

            Issued to: DigiCert Timestamp Responder
            Issued by: DigiCert Assured ID CA-1
            Expires:   Mon Oct 21 19:00:00 2024
            SHA1 hash: 614D271D9102E30169822487FDE5DE00A352B01D

Cross Certificate Chain:
    Issued to: Microsoft Code Verification Root
    Issued by: Microsoft Code Verification Root
    Expires:   Sat Nov 01 08:54:03 2025
    SHA1 hash: 8FBE4D070EF8AB1BCCAF2A9D5CCAE7282A2C66B3

        Issued to: DigiCert High Assurance EV Root CA
        Issued by: Microsoft Code Verification Root
        Expires:   Thu Apr 15 14:55:33 2021
        SHA1 hash: 2F2513AF3992DB0A3F79709FF8143B3F7BD2D143

            Issued to: DigiCert High Assurance Code Signing CA-1
            Issued by: DigiCert High Assurance EV Root CA
            Expires:   Tue Feb 10 07:00:00 2026
            SHA1 hash: E308F829DC77E80AF15EDD4151EA47C59399AB46

                Issued to: Insecure.Com LLC
                Issued by: DigiCert High Assurance Code Signing CA-1
                Expires:   Wed Jun 27 07:00:00 2018
                SHA1 hash: 67CDCA7703A01B25E6E0426072EC08B0046EB5F8

Signature Index: 1
Hash of file (sha256): CC02F1DEF532A8F323191BEAAB7A7F0B43A80E14D5EAC0A998ED7D3A0780458C

Signing Certificate Chain:
    Issued to: DigiCert High Assurance EV Root CA
    Issued by: DigiCert High Assurance EV Root CA
    Expires:   Sun Nov 09 19:00:00 2031
    SHA1 hash: 5FB7EE0633E259DBAD0C4C9AE6D38F1A61C7DC25

        Issued to: DigiCert EV Code Signing CA (SHA2)
        Issued by: DigiCert High Assurance EV Root CA
        Expires:   Sun Apr 18 07:00:00 2027
        SHA1 hash: 60EE3FC53D4BDFD1697AE5BEAE1CAB1C0F3AD4E3

            Issued to: Insecure.Com LLC
            Issued by: DigiCert EV Code Signing CA (SHA2)
            Expires:   Thu Nov 07 07:00:00 2019
            SHA1 hash: EC2AE51775F3252541B266C40528DAA77BAA072F

The signature is timestamped: Fri May 26 17:22:06 2017
Timestamp Verified by:
    Issued to: DigiCert Assured ID Root CA
    Issued by: DigiCert Assured ID Root CA
    Expires:   Sun Nov 09 19:00:00 2031
    SHA1 hash: 0563B8630D62D75ABBC8AB1E4BDFB5A899B24D43

        Issued to: DigiCert SHA2 Assured ID Timestamping CA
        Issued by: DigiCert Assured ID Root CA
        Expires:   Tue Jan 07 07:00:00 2031
        SHA1 hash: 3BA63A6E4841355772DEBEF9CDCF4D5AF353A297

            Issued to: DigiCert SHA2 Timestamp Responder
            Issued by: DigiCert SHA2 Assured ID Timestamping CA
            Expires:   Mon Jan 17 19:00:00 2028
            SHA1 hash: 400191475C98891DEBA104AF47091B5EB6D4CBCB

SignTool Error: No signature found.

Number of signatures successfully Verified: 1
Number of warnings: 0
Number of errors: 1
```

**Comment by dmiller-nmap** [comment]
@tgfruth1 Does this problem persist with later versions of Npcap? Please try again with Npcap 0.92 and let us know.

The answer on `npcap.cat` is that it is not a PE executable, so it cannot contain multiple signatures. I am not sure whether this signature is checked at all anyway, or only by Microsoft when we submit it for Attestation signing. I will change our build script to only sign it once so that there is not an error, though the signature may still not be understood by all Windows versions.

**Comment by tgfruth1** [comment]
npcap 0.92 installed and created service(s) without issue.

**Comment by djcristi** [comment]
i have the same error on version 0.93 on windows 7 x64

from the log : 
_detailprint: Failed to create the npcap service for Win7, Win8 and Win10

 {_VERIFY_FILE_SIGNATURE} 20:04:10.158
     sig:                     Key      = NPCAP.inf
     sig:                     FilePath = C:\Windows\System32\DriverStore\Temp\{2a0eee39-a35d-1817-9302-1a72b0160f03}\NPCAP.inf
     sig:                     Catalog  = C:\Windows\System32\DriverStore\Temp\{2a0eee39-a35d-1817-9302-1a72b0160f03}\npcap.cat
!    sig:                     Verifying file against specific (valid) catalog failed! (0x800b0109)
!    sig:                     Error 0x800b0109: A certificate chain processed, but terminated in a root certificate which is not trusted by the trust provider.

             {_VERIFY_FILE_SIGNATURE} 20:04:10.187
     sig:                     Key      = NPCAP.inf
     sig:                     FilePath = C:\Windows\System32\DriverStore\Temp\{2a0eee39-a35d-1817-9302-1a72b0160f03}\NPCAP.inf
     sig:                     Catalog  = C:\Windows\System32\DriverStore\Temp\{2a0eee39-a35d-1817-9302-1a72b0160f03}\npcap.cat
     sig:                     Success: File is signed in Authenticode(tm) catalog.
     sig:                     Error 0xe0000241: The INF was signed with an Authenticode(tm) catalog from a trusted publisher._

**Comment by SOlsonFAC** [comment]
I have the same error on both 0.93 and 0.94.  I was installing NMAP last week and it wanted to install NPCAP 0.93 and I got a "Failed to create the npcap services for Win7, Win8, and Win10" error.  Looking deeper, I found those same certificate errors in the setupapi.app.log file.  I'm attaching all the log files I know of (DiagReport, install.log, NPFInstall.log, setupap.app.log, setupapi.dev.log)

[DiagReport-20170830-134807.txt](https://github.com/nmap/nmap/files/1264710/DiagReport-20170830-134807.txt)
[install.log.txt](https://github.com/nmap/nmap/files/1264712/install.log.txt)
[NPFInstall.log.txt](https://github.com/nmap/nmap/files/1264713/NPFInstall.log.txt)
[setupapi.app.log.txt](https://github.com/nmap/nmap/files/1264716/setupapi.app.log.txt)
[setupapi.dev.log.txt](https://github.com/nmap/nmap/files/1264719/setupapi.dev.log.txt)

**Comment by jinalshah** [comment]
Anyone found any work arounds for this issue?

**Comment by DanyaFilatov** [comment]
Same question. Can it be solved somehow?

**Comment by dmiller-nmap** [comment]
Until we can get the next release out, you can try replacing npcap.cat with the correctly-signed one in this zip:
[npcap-cat.zip](https://github.com/nmap/nmap/files/1349676/npcap-cat.zip)

**Comment by dmiller-nmap** [comment]
Please test [Nmap 0.95](https://nmap.org/npcap/#download) and let us know if it solves this problem.

**Comment by meyerdg** [comment]
Hi. I have had this exact issue trying to install only npcap-0.97 on a Windows Server 2012 VM. It doesn't have direct network access and I'm not aware if it's missing any updates but I am able to transfer in files through management servers. 

I was able to get the install working by first installing Microsoft's Code Signing Root CA to the Local Machine in the Trusted Root Certification Authorities and Trusted Publishers store.
http://www.microsoft.com/pki/certs/MicrosoftCodeVerifRoot.crt 

Then also install the digicert assured id root ca in the same places 
https://www.digicert.com/CACerts/DigiCertAssuredIDRootCA.crt

Then install the digicert assured id ca-1 to the local machine in the Intermediate Certification Authorities store.
https://www.digicert.com/CACerts/DigiCertAssuredIDCA-1.crt

Hope that helps. Would be better if I knew this machine was fully up to date but that isn't under my control so this work-around helped. 

I also installed certificates as described above by hsluoyz on May 8, 2017
